File: sort-mapper-example-java/src/main/java/kh/farrukh/sortmapperexamplejava/course/CourseEntity.java
package kh.farrukh.sortmapperexamplejava.course;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EntityListeners;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.SequenceGenerator;
import jakarta.persistence.Table;
import kh.farrukh.sortmapperexamplejava.common.audit.AuditableEntity;
import kh.farrukh.sortmapperexamplejava.teacher.TeacherEntity;
import lombok.Getter;
import lombok.Setter;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

@Getter
@Setter
@Entity
@Table(name = "courses")
@EntityListeners(AuditingEntityListener.class)
public class CourseEntity extends AuditableEntity {

  private static final String COURSE_ID_GENERATOR = "course_id_generator";
  private static final String COURSE_ID_SEQUENCE = "course_id_sequence";

  @Id
  @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = COURSE_ID_GENERATOR)
  @SequenceGenerator(name = COURSE_ID_GENERATOR, sequenceName = COURSE_ID_SEQUENCE, allocationSize = 1)
  private Long id;

  @Column(name = "title", nullable = false, length = 50)
  private String title;

  @Column(name = "description", nullable = false, length = 200)
  private String description;

  @Column(name = "slogan", length = 100)
  private String slogan;

  @ManyToOne(optional = false)
  @JoinColumn(name = "teacher_id", nullable = false)
  private TeacherEntity teacher;

  @Column(name = "duration_month", nullable = false)
  private Integer durationMonth;

  @Column(name = "price", nullable = false)
  private Double price;

}

File: sort-mapper-example/src/main/kotlin/kh/farrukh/sortmapperexample/course/CourseEntity.kt
package kh.farrukh.sortmapperexample.course

import jakarta.annotation.Nonnull
import jakarta.persistence.*
import kh.farrukh.sortmapperexample.common.audit.AuditableEntity
import kh.farrukh.sortmapperexample.teacher.TeacherEntity
import org.jetbrains.annotations.NotNull
import org.springframework.data.jpa.domain.support.AuditingEntityListener
import java.time.LocalDateTime

@Entity
@Table(name = "courses")
@EntityListeners(AuditingEntityListener::class)
class CourseEntity(
    @Id
    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = COURSE_ID_GENERATOR)
    @SequenceGenerator(
        name = COURSE_ID_GENERATOR,
        sequenceName = COURSE_ID_SEQUENCE,
        allocationSize = 1
    )
    var id: Long? = null,
    @Column(name = "title", nullable = false, length = 50)
    var title: String,
    @Column(name = "description", nullable = false, length = 200)
    var description: String,
    @Column(name = "slogan", length = 100)
    var slogan: String? = null,
    @ManyToOne(optional = false)
    @NotNull
    @JoinColumn(name = "teacher_id", nullable = false)
    var teacher: TeacherEntity,
    @Column(name = "duration_month", nullable = false)
    var durationMonth: Int,
    @Column(name = "price", nullable = false)
    var price: Double
) : AuditableEntity(
    createdAt = LocalDateTime.now(),
    updatedAt = LocalDateTime.now()
) {

    companion object {
        const val COURSE_ID_GENERATOR = "course_id_generator"
        const val COURSE_ID_SEQUENCE = "course_id_sequence"
    }

}


